File: ai engine/utils/report_pdf.py
"""Server-side PDF generation for daily reports using ReportLab.

Produces a styled A4 PDF matching the frontend design:
  - Dark header banner with title + date
  - Coloured stat cards
  - Executive Summary
  - Top Posts
  - Karma Leaderboard table
  - Moderation Overview
  - Footer on every page
"""

from __future__ import annotations

import io
import logging
from datetime import datetime
from typing import Any

from reportlab.lib import colors
from reportlab.lib.pagesizes import A4
from reportlab.lib.units import mm
from reportlab.pdfgen.canvas import Canvas
from reportlab.lib.styles import getSampleStyleSheet, ParagraphStyle
from reportlab.platypus import Paragraph
from reportlab.lib.enums import TA_LEFT

logger = logging.getLogger(__name__)

# ── Colour palette ───────────────────────────────────────────────────────────
_PURPLE = colors.HexColor("#8B5CF6")
_PINK = colors.HexColor("#EC4899")
_GREEN = colors.HexColor("#22C55E")
_AMBER = colors.HexColor("#F59E0B")
_RED = colors.HexColor("#EF4444")
_DARK = colors.HexColor("#0F172A")
_TEXT = colors.HexColor("#1E293B")
_MUTED = colors.HexColor("#64748B")
_CARD_BG = colors.HexColor("#F8FAFC")
_WHITE = colors.white

PAGE_W, PAGE_H = A4  # 595.28 x 841.89 pts
MARGIN = 50


def _safe(text: str) -> str:
    """Replace problematic Unicode chars with ASCII equivalents."""
    return (
        text
        .replace("\u2018", "'").replace("\u2019", "'").replace("\u201A", "'")
        .replace("\u201C", '"').replace("\u201D", '"').replace("\u201E", '"')
        .replace("\u2013", "-").replace("\u2014", "--")
        .replace("\u2026", "...")
        .replace("\u25C9", "*").replace("\u2022", "*").replace("\u25CF", "*")
        .replace("\u2713", "v").replace("\u2714", "v")
    )


def generate_report_pdf(report: dict[str, Any]) -> bytes:
    """Return raw PDF bytes for the given report dict."""
    buf = io.BytesIO()
    c = Canvas(buf, pagesize=A4)
    content_w = PAGE_W - MARGIN * 2

    y = PAGE_H  # current vertical position (top-down)

    # ═════════════════════════════════════════════════════════════════════════
    # HEADER BANNER
    # ═════════════════════════════════════════════════════════════════════════
    header_h = 110
    # Dark background
    c.setFillColor(_DARK)
    c.rect(0, PAGE_H - header_h, PAGE_W, header_h, fill=True, stroke=False)
    # Accent bar
    c.setFillColor(_PURPLE)
    c.rect(0, PAGE_H - header_h, PAGE_W, 8, fill=True, stroke=False)
    c.setFillColor(_PINK)
    c.rect(0, PAGE_H - header_h, PAGE_W * 0.35, 4, fill=True, stroke=False)

    # Title
    c.setFillColor(_WHITE)
    c.setFont("Helvetica-Bold", 14)
    c.drawString(MARGIN, PAGE_H - 30, "AI OBSERVATORY")
    c.setFont("Helvetica", 9)
    c.setFillColor(colors.HexColor("#C8C8DC"))
    c.drawString(MARGIN, PAGE_H - 42, "DAILY INTELLIGENCE REPORT")

    # Date (right-aligned)
    report_date = report.get("report_date", "")
    try:
        date_display = datetime.strptime(report_date, "%Y-%m-%d").strftime("%A, %B %d, %Y")
    except Exception:
        date_display = report_date

    c.setFillColor(_WHITE)
    c.setFont("Helvetica-Bold", 12)
    c.drawRightString(PAGE_W - MARGIN, PAGE_H - 30, date_display)

    created_at = report.get("created_at", "")
    if created_at:
        try:
            gen_dt = datetime.fromisoformat(created_at.replace("Z", "+00:00"))
            gen_display = gen_dt.strftime("Generated: %A, %B %d, %Y at %I:%M %p")
        except Exception:
            gen_display = f"Generated: {created_at}"
        c.setFont("Helvetica", 7)
        c.setFillColor(colors.HexColor("#B4B4C8"))
        c.drawRightString(PAGE_W - MARGIN, PAGE_H - 42, gen_display)

    # Headline
    headline = _safe(report.get("headline", f"Daily Report -- {report_date}"))
    c.setFillColor(_WHITE)
    c.setFont("Helvetica-Bold", 16)
    # Wrap headline if too long
    headline_lines = _wrap_text(c, headline, content_w, "Helvetica-Bold", 16)
    hy = PAGE_H - 65
    for hl in headline_lines[:2]:
        c.drawString(MARGIN, hy, hl)
        hy -= 20

    y = PAGE_H - header_h - 15

    # ═════════════════════════════════════════════════════════════════════════
    # STAT CARDS
    # ═════════════════════════════════════════════════════════════════════════
    stats = [
        ("NEWS INGESTED", report.get("news_count", 0), _PURPLE),
        ("AGENT POSTS", report.get("post_count", 0), _PINK),
        ("TOTAL ACTIVITY", report.get("agent_findings_count", 0), _GREEN),
    ]

    card_w = (content_w - 20) / 3
    card_h = 55
    for i, (label, value, colour) in enumerate(stats):
        cx = MARGIN + i * (card_w + 10)
        cy = y - card_h

        # Card bg
        c.setFillColor(_CARD_BG)
        c.roundRect(cx, cy, card_w, card_h, 6, fill=True, stroke=False)

        # Colour bar at top
        c.setFillColor(colour)
        c.rect(cx, cy + card_h - 6, card_w, 6, fill=True, stroke=False)

        # Value
        c.setFillColor(colour)
        c.setFont("Helvetica-Bold", 22)
        c.drawCentredString(cx + card_w / 2, cy + 20, str(value))

        # Label
        c.setFillColor(_MUTED)
        c.setFont("Helvetica", 7)
        c.drawCentredString(cx + card_w / 2, cy + 8, label)

    y -= card_h + 20

    # ═════════════════════════════════════════════════════════════════════════
    # EXECUTIVE SUMMARY
    # ═════════════════════════════════════════════════════════════════════════
    summary = report.get("summary", "")
    if summary:
        y = _check_page(c, y, 60)

        # Section heading
        c.setFillColor(_PURPLE)
        c.roundRect(MARGIN, y - 12, 4, 14, 2, fill=True, stroke=False)
        c.setFillColor(_DARK)
        c.setFont("Helvetica-Bold", 13)
        c.drawString(MARGIN + 10, y - 8, "Executive Summary")
        y -= 25

        # Body paragraphs
        c.setFont("Helvetica", 10)
        c.setFillColor(_TEXT)
        for para in _safe(summary).split("\n"):
            para = para.strip()
            if not para:
                continue
            lines = _wrap_text(c, para, content_w - 8, "Helvetica", 10)
            for line in lines:
                y = _check_page(c, y, 14)
                c.drawString(MARGIN + 4, y, line)
                y -= 14
            y -= 6

    # ═════════════════════════════════════════════════════════════════════════
    # TOP POSTS
    # ═════════════════════════════════════════════════════════════════════════
    top_posts = report.get("top_posts", [])
    if top_posts:
        y -= 8
        y = _check_page(c, y, 40)

        c.setFillColor(_PINK)
        c.roundRect(MARGIN, y - 12, 4, 14, 2, fill=True, stroke=False)
        c.setFillColor(_DARK)
        c.setFont("Helvetica-Bold", 13)
        c.drawString(MARGIN + 10, y - 8, "Top Posts")
        y -= 25

        for idx, post in enumerate(top_posts[:5]):
            y = _check_page(c, y, 50)

            agent_name = _safe(str(post.get("agent_name", "Unknown")))
            body = _safe(str(post.get("body", "")))
            upvotes = post.get("upvote_count", post.get("upvotes", 0))
            downvotes = post.get("downvote_count", post.get("downvotes", 0))

            # Rank + agent name
            c.setFont("Helvetica-Bold", 10)
            c.setFillColor(_DARK)
            c.drawString(MARGIN + 4, y, f"{idx + 1}. {agent_name}")

            # Vote badges
            vote_text = f"+{upvotes} / -{downvotes}"
            vx = MARGIN + 8 + c.stringWidth(f"{idx + 1}. {agent_name}", "Helvetica-Bold", 10)
            c.setFillColor(_GREEN)
            c.setFont("Helvetica-Bold", 8)
            c.drawString(vx + 6, y + 1, vote_text)
            y -= 14

            # Post body (truncated)
            c.setFont("Helvetica", 9)
            c.setFillColor(_TEXT)
            body_lines = _wrap_text(c, body[:300], content_w - 16, "Helvetica", 9)
            for bl in body_lines[:4]:
                y = _check_page(c, y, 12)
                c.drawString(MARGIN + 12, y, bl)
                y -= 12
            y -= 8

    # ═════════════════════════════════════════════════════════════════════════
    # KARMA LEADERBOARD
    # ═════════════════════════════════════════════════════════════════════════
    karma_lb = report.get("karma_leaderboard", [])
    if karma_lb:
        y -= 8
        y = _check_page(c, y, 50)

        c.setFillColor(_GREEN)
        c.roundRect(MARGIN, y - 12, 4, 14, 2, fill=True, stroke=False)
        c.setFillColor(_DARK)
        c.setFont("Helvetica-Bold", 13)
        c.drawString(MARGIN + 10, y - 8, "Karma Leaderboard")
        y -= 22

        # Table header
        row_h = 20
        c.setFillColor(_DARK)
        c.roundRect(MARGIN, y - row_h + 4, content_w, row_h, 4, fill=True, stroke=False)
        c.setFillColor(_WHITE)
        c.setFont("Helvetica-Bold", 8)
        c.drawString(MARGIN + 8, y - 8, "RANK")
        c.drawString(MARGIN + 50, y - 8, "AGENT")
        c.drawString(MARGIN + content_w - 100, y - 8, "KARMA")
        c.drawString(MARGIN + content_w - 50, y - 8, "STATUS")
        y -= row_h + 2

        for idx, agent in enumerate(karma_lb[:10]):
            y = _check_page(c, y, 18)

            bg = _CARD_BG if idx % 2 == 0 else _WHITE
            c.setFillColor(bg)
            c.roundRect(MARGIN, y - 14, content_w, 18, 2, fill=True, stroke=False)

            c.setFont("Helvetica-Bold", 9)
            c.setFillColor(_DARK if idx < 3 else _MUTED)
            c.drawString(MARGIN + 8, y - 8, f"#{idx + 1}")

            c.setFont("Helvetica", 9)
            c.setFillColor(_TEXT)
            c.drawString(MARGIN + 50, y - 8, _safe(str(agent.get("agent_name", agent.get("name", "?")))))

            c.setFont("Helvetica-Bold", 9)
            c.setFillColor(_PURPLE)
            c.drawString(MARGIN + content_w - 100, y - 8, str(agent.get("karma", 0)))

            is_verified = agent.get("is_verified", False)
            if is_verified:
                c.setFillColor(_GREEN)
                c.roundRect(MARGIN + content_w - 58, y - 12, 40, 14, 4, fill=True, stroke=False)
                c.setFillColor(_WHITE)
                c.setFont("Helvetica-Bold", 7)
                c.drawCentredString(MARGIN + content_w - 38, y - 8, "VERIFIED")
            else:
                c.setFillColor(_MUTED)
                c.setFont("Helvetica", 7)
                c.drawCentredString(MARGIN + content_w - 38, y - 8, "--")

            y -= 20

    # ═════════════════════════════════════════════════════════════════════════
    # MODERATION STATS
    # ═════════════════════════════════════════════════════════════════════════
    mod = report.get("moderation_stats", {})
    if mod and any(mod.get(k) for k in ("reviewed", "approved", "flagged", "rejected")):
        y -= 12
        y = _check_page(c, y, 60)

        c.setFillColor(_AMBER)
        c.roundRect(MARGIN, y - 12, 4, 14, 2, fill=True, stroke=False)
        c.setFillColor(_DARK)
        c.setFont("Helvetica-Bold", 13)
        c.drawString(MARGIN + 10, y - 8, "Moderation Overview")
        y -= 25

        mod_stats = [
            ("Reviewed", mod.get("reviewed", 0), _PURPLE),
            ("Approved", mod.get("approved", 0), _GREEN),
            ("Flagged", mod.get("flagged", 0), _AMBER),
            ("Rejected", mod.get("rejected", 0), _RED),
        ]
        mod_w = (content_w - 30) / 4
        mod_h = 45
        for i, (label, val, colour) in enumerate(mod_stats):
            mx = MARGIN + i * (mod_w + 10)
            my = y - mod_h
            c.setFillColor(_CARD_BG)
            c.roundRect(mx, my, mod_w, mod_h, 4, fill=True, stroke=False)

            c.setFillColor(colour)
            c.setFont("Helvetica-Bold", 18)
            c.drawCentredString(mx + mod_w / 2, my + 22, str(val))

            c.setFillColor(_MUTED)
            c.setFont("Helvetica", 7)
            c.drawCentredString(mx + mod_w / 2, my + 8, label.upper())

        y -= mod_h + 10

    # ═════════════════════════════════════════════════════════════════════════
    # FOOTER (all pages)
    # ═════════════════════════════════════════════════════════════════════════
    total_pages = c.getPageNumber()
    _draw_footer(c, total_pages)
    c.save()

    # Add footers to all pages (re-iterate is not needed — we drew inline)
    pdf_bytes = buf.getvalue()
    buf.close()
    return pdf_bytes


# ── Helpers ──────────────────────────────────────────────────────────────────

def _wrap_text(c: Canvas, text: str, max_width: float, font: str, size: float) -> list[str]:
    """Word-wrap *text* to fit within *max_width* using the given font."""
    words = text.split()
    lines: list[str] = []
    current = ""
    for word in words:
        test = f"{current} {word}".strip()
        if c.stringWidth(test, font, size) <= max_width:
            current = test
        else:
            if current:
                lines.append(current)
            current = word
    if current:
        lines.append(current)
    return lines or [""]


def _check_page(c: Canvas, y: float, needed: float) -> float:
    """If there's not enough room, start a new page and return fresh y."""
    if y - needed < 40:
        _draw_footer(c, c.getPageNumber())
        c.showPage()
        return PAGE_H - MARGIN
    return y


def _draw_footer(c: Canvas, page_num: int) -> None:
    """Draw a footer line at the bottom of the current page."""
    fy = 30
    c.setStrokeColor(colors.HexColor("#E2E8F0"))
    c.setLineWidth(0.3)
    c.line(MARGIN, fy + 4, PAGE_W - MARGIN, fy + 4)
    c.setFillColor(_MUTED)
    c.setFont("Helvetica", 7)
    c.drawString(MARGIN, fy - 4, "AI Observatory -- Automated Daily Intelligence Report")
    c.drawRightString(PAGE_W - MARGIN, fy - 4, f"Page {page_num}")

File: frontend/Agent Watch/src/lib/report-pdf.ts
/**
 * Generates a beautiful, colorful PDF from a daily report object using jsPDF.
 *
 * NOTE: jsPDF's built-in fonts (Helvetica, Courier, Times) only support
 * WinAnsi / Latin-1 characters.  We sanitise all text through `safe()`
 * to replace Unicode dashes, quotes, bullets etc. with ASCII equivalents.
 */
import jsPDF from "jspdf";

// ── Color palette ───────────────────────────────────────────────────────────
const COLORS = {
  primary: [139, 92, 246] as [number, number, number],       // purple-500
  primaryDark: [109, 40, 217] as [number, number, number],   // purple-600
  accent: [236, 72, 153] as [number, number, number],        // pink-500
  success: [34, 197, 94] as [number, number, number],        // green-500
  warning: [245, 158, 11] as [number, number, number],       // amber-500
  danger: [239, 68, 68] as [number, number, number],         // red-500
  dark: [15, 23, 42] as [number, number, number],            // slate-900
  text: [30, 41, 59] as [number, number, number],            // slate-800
  muted: [100, 116, 139] as [number, number, number],        // slate-500
  light: [241, 245, 249] as [number, number, number],        // slate-100
  white: [255, 255, 255] as [number, number, number],
  card: [248, 250, 252] as [number, number, number],         // slate-50
};

// ── Sanitise text for jsPDF (Latin-1 only) ──────────────────────────────────
function safe(input: string): string {
  return input
    // Smart quotes
    .replace(/[\u2018\u2019\u201A]/g, "'")
    .replace(/[\u201C\u201D\u201E]/g, '"')
    // Dashes
    .replace(/\u2013/g, "-")   // en-dash
    .replace(/\u2014/g, "--")  // em-dash
    // Ellipsis
    .replace(/\u2026/g, "...")
    // Bullets / symbols
    .replace(/\u25C9/g, "*")   // ◉
    .replace(/\u2022/g, "*")   // •
    .replace(/\u25CF/g, "*")   // ●
    .replace(/\u2713/g, "v")   // ✓
    .replace(/\u2714/g, "v")   // ✔
    // Arrows
    .replace(/\u2192/g, "->")
    .replace(/\u2190/g, "<-")
    // Math / misc
    .replace(/\u00D7/g, "x")   // ×
    .replace(/\u2265/g, ">=")  // ≥
    .replace(/\u2264/g, "<=")  // ≤
    // Remove any remaining non-Latin-1 chars (above U+00FF) that would garble
    .replace(/[^\x00-\xFF]/g, "");
}

interface ReportData {
  id: string;
  report_date: string;
  created_at?: string;
  headline?: string;
  summary?: string;
  news_count?: number;
  post_count?: number;
  agent_findings_count?: number;
  top_posts?: Array<{
    body: string;
    agent_name: string;
    upvote_count: number;
    downvote_count: number;
  }>;
  karma_leaderboard?: Array<{
    agent_name: string;
    karma: number;
    is_verified: boolean;
  }>;
  moderation_stats?: {
    reviewed?: number;
    approved?: number;
    flagged?: number;
    rejected?: number;
  };
}

function formatDate(dateStr: string): string {
  const d = new Date(dateStr + "T00:00:00");
  return d.toLocaleDateString("en-US", {
    weekday: "long",
    year: "numeric",
    month: "long",
    day: "numeric",
  });
}

function formatDateTime(dateStr: string): string {
  const d = new Date(dateStr);
  return d.toLocaleString("en-US", {
    weekday: "long",
    year: "numeric",
    month: "long",
    day: "numeric",
    hour: "2-digit",
    minute: "2-digit",
    timeZoneName: "short",
  });
}

export function downloadReportPdf(report: ReportData) {
  const doc = new jsPDF({ orientation: "portrait", unit: "mm", format: "a4" });
  const pageW = doc.internal.pageSize.getWidth();
  const pageH = doc.internal.pageSize.getHeight();
  const margin = 18;
  const contentW = pageW - margin * 2;
  let y = 0;

  // ── Helper: check if we need a new page ─────────────────────────────────
  const checkPage = (needed: number) => {
    if (y + needed > pageH - 20) {
      doc.addPage();
      y = margin;
    }
  };

  // ── Helper: draw rounded rect ───────────────────────────────────────────
  const roundedRect = (
    x: number,
    yy: number,
    w: number,
    h: number,
    r: number,
    fill: [number, number, number],
  ) => {
    doc.setFillColor(...fill);
    doc.roundedRect(x, yy, w, h, r, r, "F");
  };

  // ═══════════════════════════════════════════════════════════════════════════
  // HEADER BANNER
  // ═══════════════════════════════════════════════════════════════════════════

  // Gradient-like header (two overlapping rects)
  doc.setFillColor(...COLORS.dark);
  doc.rect(0, 0, pageW, 52, "F");
  doc.setFillColor(...COLORS.primary);
  doc.rect(0, 46, pageW, 6, "F");

  // Decorative accent bar
  doc.setFillColor(...COLORS.accent);
  doc.rect(0, 50, pageW * 0.35, 2, "F");
  doc.setFillColor(...COLORS.primary);
  doc.rect(pageW * 0.35, 50, pageW * 0.65, 2, "F");

  // Logo / Title — a small purple circle as the "logo"
  doc.setFillColor(...COLORS.primary);
  doc.circle(margin + 3, 12, 3, "F");
  doc.setFillColor(...COLORS.accent);
  doc.circle(margin + 3, 12, 1.5, "F");

  doc.setTextColor(...COLORS.white);
  doc.setFont("helvetica", "bold");
  doc.setFontSize(10);
  doc.text("AI OBSERVATORY", margin + 9, 14);

  doc.setFontSize(8);
  doc.setFont("helvetica", "normal");
  doc.setTextColor(200, 200, 220);
  doc.text("DAILY INTELLIGENCE REPORT", margin + 9, 20);

  // Report date (right-aligned)
  doc.setTextColor(...COLORS.white);
  doc.setFont("helvetica", "bold");
  doc.setFontSize(11);
  doc.text(safe(formatDate(report.report_date)), pageW - margin, 14, {
    align: "right",
  });

  if (report.created_at) {
    doc.setFontSize(7);
    doc.setFont("helvetica", "normal");
    doc.setTextColor(180, 180, 200);
    doc.text(
      safe(`Generated: ${formatDateTime(report.created_at)}`),
      pageW - margin,
      20,
      { align: "right" },
    );
  }

  // Headline
  doc.setTextColor(...COLORS.white);
  doc.setFont("helvetica", "bold");
  doc.setFontSize(15);
  const headlineText = safe(
    report.headline || `Daily Report -- ${report.report_date}`,
  );
  const headlineLines = doc.splitTextToSize(headlineText, contentW);
  doc.text(headlineLines, margin, 32);

  y = 58;

  // ═══════════════════════════════════════════════════════════════════════════
  // STATS ROW
  // ═══════════════════════════════════════════════════════════════════════════

  const stats = [
    {
      label: "News Ingested",
      value: String(report.news_count ?? 0),
      color: COLORS.primary,
    },
    {
      label: "Agent Posts",
      value: String(report.post_count ?? 0),
      color: COLORS.accent,
    },
    {
      label: "Total Activity",
      value: String(report.agent_findings_count ?? 0),
      color: COLORS.success,
    },
  ];

  const statW = (contentW - 6) / 3;
  stats.forEach((s, i) => {
    const sx = margin + i * (statW + 3);
    roundedRect(sx, y, statW, 22, 3, COLORS.card);

    // Color indicator bar at top
    doc.setFillColor(...s.color);
    doc.roundedRect(sx, y, statW, 3, 3, 3, "F");
    doc.setFillColor(...COLORS.card);
    doc.rect(sx, y + 2, statW, 2, "F");

    doc.setFont("helvetica", "bold");
    doc.setFontSize(18);
    doc.setTextColor(...s.color);
    doc.text(s.value, sx + statW / 2, y + 13, { align: "center" });

    doc.setFont("helvetica", "normal");
    doc.setFontSize(7);
    doc.setTextColor(...COLORS.muted);
    doc.text(s.label.toUpperCase(), sx + statW / 2, y + 19, {
      align: "center",
    });
  });

  y += 28;

  // ═══════════════════════════════════════════════════════════════════════════
  // SUMMARY
  // ═══════════════════════════════════════════════════════════════════════════

  if (report.summary) {
    checkPage(40);

    // Section heading
    doc.setFillColor(...COLORS.primary);
    doc.roundedRect(margin, y, 3, 10, 1.5, 1.5, "F");
    doc.setFont("helvetica", "bold");
    doc.setFontSize(12);
    doc.setTextColor(...COLORS.dark);
    doc.text("Executive Summary", margin + 7, y + 7);
    y += 15;

    // Summary body — split by newlines for paragraphs
    doc.setFont("helvetica", "normal");
    doc.setFontSize(9.5);
    doc.setTextColor(...COLORS.text);
    const paragraphs = safe(report.summary)
      .split("\n")
      .filter((p) => p.trim());
    paragraphs.forEach((para) => {
      checkPage(20);
      const lines = doc.splitTextToSize(para.trim(), contentW - 4);
      doc.text(lines, margin + 2, y);
      y += lines.length * 4.5 + 4;
    });

    y += 2;
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // TOP POSTS
  // ═══════════════════════════════════════════════════════════════════════════

  if (report.top_posts && report.top_posts.length > 0) {
    checkPage(30);

    // Section heading
    doc.setFillColor(...COLORS.accent);
    doc.roundedRect(margin, y, 3, 10, 1.5, 1.5, "F");
    doc.setFont("helvetica", "bold");
    doc.setFontSize(12);
    doc.setTextColor(...COLORS.dark);
    doc.text("Top Posts", margin + 7, y + 7);
    y += 15;

    report.top_posts.forEach((post, idx) => {
      checkPage(25);

      const safeBody = safe(post.body || "");
      const safeName = safe(post.agent_name || "Unknown");

      // Card background
      const bodyLines = doc.splitTextToSize(safeBody, contentW - 28);
      const cardH = Math.max(18, bodyLines.length * 4 + 14);
      roundedRect(margin, y, contentW, cardH, 2, COLORS.card);

      // Rank badge
      const badgeColor =
        idx === 0 ? COLORS.primary : idx === 1 ? COLORS.accent : COLORS.muted;
      doc.setFillColor(...badgeColor);
      doc.circle(margin + 6, y + 7, 4, "F");
      doc.setTextColor(...COLORS.white);
      doc.setFont("helvetica", "bold");
      doc.setFontSize(8);
      doc.text(String(idx + 1), margin + 6, y + 8.5, { align: "center" });

      // Agent name + votes
      doc.setTextColor(...COLORS.dark);
      doc.setFont("helvetica", "bold");
      doc.setFontSize(9);
      doc.text(safeName, margin + 14, y + 7);

      // Vote badges
      const voteX = margin + 14 + doc.getTextWidth(safeName) + 4;
      // Upvotes
      doc.setFillColor(...COLORS.success);
      doc.roundedRect(voteX, y + 3, 14, 6, 2, 2, "F");
      doc.setTextColor(...COLORS.white);
      doc.setFontSize(7);
      doc.setFont("helvetica", "bold");
      doc.text(`+${post.upvote_count}`, voteX + 7, y + 7.5, {
        align: "center",
      });
      // Downvotes
      if (post.downvote_count > 0) {
        doc.setFillColor(...COLORS.danger);
        doc.roundedRect(voteX + 16, y + 3, 14, 6, 2, 2, "F");
        doc.setTextColor(...COLORS.white);
        doc.text(`-${post.downvote_count}`, voteX + 23, y + 7.5, {
          align: "center",
        });
      }

      // Post body
      doc.setFont("helvetica", "normal");
      doc.setFontSize(8.5);
      doc.setTextColor(...COLORS.text);
      doc.text(bodyLines, margin + 14, y + 13);

      y += cardH + 3;
    });

    y += 2;
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // KARMA LEADERBOARD
  // ═══════════════════════════════════════════════════════════════════════════

  if (report.karma_leaderboard && report.karma_leaderboard.length > 0) {
    checkPage(30);

    // Section heading
    doc.setFillColor(...COLORS.success);
    doc.roundedRect(margin, y, 3, 10, 1.5, 1.5, "F");
    doc.setFont("helvetica", "bold");
    doc.setFontSize(12);
    doc.setTextColor(...COLORS.dark);
    doc.text("Karma Leaderboard", margin + 7, y + 7);
    y += 14;

    // Table header
    roundedRect(margin, y, contentW, 8, 2, COLORS.dark);
    doc.setTextColor(...COLORS.white);
    doc.setFont("helvetica", "bold");
    doc.setFontSize(7.5);
    doc.text("RANK", margin + 6, y + 5.5);
    doc.text("AGENT", margin + 22, y + 5.5);
    doc.text("KARMA", margin + contentW - 40, y + 5.5);
    doc.text("STATUS", margin + contentW - 16, y + 5.5);
    y += 9;

    report.karma_leaderboard.forEach((agent, idx) => {
      checkPage(10);
      const rowColor = idx % 2 === 0 ? COLORS.card : COLORS.white;
      roundedRect(margin, y, contentW, 8, 1, rowColor);

      // Rank
      doc.setFont("helvetica", "bold");
      doc.setFontSize(8);
      if (idx < 3) {
        doc.setTextColor(...COLORS.dark);
      } else {
        doc.setTextColor(...COLORS.muted);
      }
      doc.text(`#${idx + 1}`, margin + 6, y + 5.5);

      // Agent name
      doc.setFont("helvetica", "normal");
      doc.setFontSize(8.5);
      doc.setTextColor(...COLORS.text);
      doc.text(safe(agent.agent_name), margin + 22, y + 5.5);

      // Karma
      doc.setFont("helvetica", "bold");
      doc.setTextColor(...COLORS.primary);
      doc.text(String(agent.karma), margin + contentW - 40, y + 5.5);

      // Verified badge
      if (agent.is_verified) {
        doc.setFillColor(...COLORS.success);
        doc.roundedRect(margin + contentW - 22, y + 1.5, 18, 5, 2, 2, "F");
        doc.setTextColor(...COLORS.white);
        doc.setFontSize(6);
        doc.text("VERIFIED", margin + contentW - 13, y + 5, {
          align: "center",
        });
      } else {
        doc.setTextColor(...COLORS.muted);
        doc.setFontSize(6);
        doc.text("--", margin + contentW - 13, y + 5, { align: "center" });
      }

      y += 9;
    });

    y += 4;
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // MODERATION STATS
  // ═══════════════════════════════════════════════════════════════════════════

  const mod = report.moderation_stats;
  if (mod && (mod.reviewed || mod.approved || mod.flagged || mod.rejected)) {
    checkPage(30);

    doc.setFillColor(...COLORS.warning);
    doc.roundedRect(margin, y, 3, 10, 1.5, 1.5, "F");
    doc.setFont("helvetica", "bold");
    doc.setFontSize(12);
    doc.setTextColor(...COLORS.dark);
    doc.text("Moderation Overview", margin + 7, y + 7);
    y += 15;

    const modStats = [
      { label: "Reviewed", value: mod.reviewed ?? 0, color: COLORS.primary },
      { label: "Approved", value: mod.approved ?? 0, color: COLORS.success },
      { label: "Flagged", value: mod.flagged ?? 0, color: COLORS.warning },
      { label: "Rejected", value: mod.rejected ?? 0, color: COLORS.danger },
    ];

    const modW = (contentW - 9) / 4;
    modStats.forEach((s, i) => {
      const mx = margin + i * (modW + 3);
      roundedRect(mx, y, modW, 18, 2, COLORS.card);

      doc.setFont("helvetica", "bold");
      doc.setFontSize(16);
      doc.setTextColor(...s.color);
      doc.text(String(s.value), mx + modW / 2, y + 10, { align: "center" });

      doc.setFont("helvetica", "normal");
      doc.setFontSize(6.5);
      doc.setTextColor(...COLORS.muted);
      doc.text(s.label.toUpperCase(), mx + modW / 2, y + 15, {
        align: "center",
      });
    });

    y += 24;
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // FOOTER (on every page)
  // ═══════════════════════════════════════════════════════════════════════════

  const totalPages = doc.getNumberOfPages();
  for (let p = 1; p <= totalPages; p++) {
    doc.setPage(p);
    const footerY = pageH - 12;
    doc.setDrawColor(...COLORS.light);
    doc.setLineWidth(0.3);
    doc.line(margin, footerY - 3, pageW - margin, footerY - 3);

    doc.setFontSize(7);
    doc.setFont("helvetica", "normal");
    doc.setTextColor(...COLORS.muted);
    doc.text(
      "AI Observatory -- Automated Daily Intelligence Report",
      margin,
      footerY,
    );
    doc.text(`Page ${p} of ${totalPages}`, pageW - margin, footerY, {
      align: "right",
    });
  }

  // ── Save ──────────────────────────────────────────────────────────────────
  doc.save(`observatory-report-${report.report_date}.pdf`);
}
